security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -269,21 +269,27 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
 			sdsMetrics.pendingPushPerConn.WithLabelValues(generateResourcePerConnLabel(resourceName, conID)).Inc()
 			sdsMetrics.totalActiveConn.Inc()
 
+			if discReq.ErrorDetail != nil {
+				sdsMetrics.totalSecretUpdateFailureCounts.Inc()
+			}
 			// When nodeagent receives StreamSecrets request, if there is cached secret which matches
 			// request's <token, resourceName, Version>, then this request is a confirmation request.
 			// nodeagent stops sending response to envoy in this case.
 			if discReq.VersionInfo != "" && s.st.SecretExist(conID, resourceName, token, discReq.VersionInfo) {
-				sdsServiceLog.Debugf("%s received SDS ACK from proxy %q, versionInfo %q\n",
-					conIDresourceNamePrefix, discReq.Node.Id, discReq.VersionInfo)
+				sdsServiceLog.Debugf("%s received SDS ACK from proxy %q, version info %q, "+
+					"error details %s\n", conIDresourceNamePrefix, discReq.Node.Id, discReq.VersionInfo,
+					discReq.ErrorDetail.GoString())
 				continue
 			}
 
 			if firstRequestFlag {
-				sdsServiceLog.Debugf("%s received first SDS request from proxy %q, versionInfo %q\n",
-					conIDresourceNamePrefix, discReq.Node.Id, discReq.VersionInfo)
+				sdsServiceLog.Debugf("%s received first SDS request from proxy %q, version info "+
+					"%q, error details %s\n", conIDresourceNamePrefix, discReq.Node.Id, discReq.VersionInfo,
+					discReq.ErrorDetail.GoString())
```

Comment:
could ```ErrorDetail```  be set for none-(ack/nack) requests ?

Reply:
Because Citadel agent closes a stream in certain conditions, and Envoy SDS will reconnect once that stream is closed. After stream is reconnected, Envoy sends SDS request which has version info and nonce (if previous SDS request is responded with valid secret), and error detail (if previous update is failed). Citadel agent does not differentiate whether a new connection is created by a new SDS client, or is re-connected by an existing SDS client.